lib/core/src/Cardano/Wallet/Primitive/Types/StateDeltaSeq.hs

```diff
@@ -0,0 +1,541 @@
+-- the 'StateDeltaSeq' type does provide several functions for constructing
+-- sequences that are valid, and verifying that existing sequences are valid.
+--
+-- Basic usage:
+--
+--    - Use 'fromStateDeltas' to construct a sequence.
+--    - Use 'applyDelta' to extend a sequence.
+--    - Use 'isValid' to verify a sequence.
+--    - Use 'toTransitionList' to list all transitions of a sequence.
+--
+module Cardano.Wallet.Primitive.Types.StateDeltaSeq
+    (
+    -- * Types
+      StateDeltaSeq
+
+    -- * Constructors
+    , fromState
+    , fromStateDeltas
+    , fromStateDeltasUnchecked
+
+    -- * Indicators
+    , isPrefixOf
+    , isSuffixOf
+    , isValid
+    , isValidM
+
+    -- * Conversions
+    , toDeltaList
+    , toStateList
+    , toTransitionList
+
+    -- * Views
+    , headState
+    , lastState
+
+    -- * Maps
+    , mapDeltas
+    , mapStates
+    , mapStatesDeltas
+
+    -- * Counts
+    , countTransitions
+    , countTransitionsWhere
+    , countEmptyTransitions
+    , countEmptyTransitionsWhere
+
+    -- * Extension
+    , applyDelta
+    , applyDeltas
+    , applyDeltaM
+    , applyDeltasM
+
+    -- * Shrinking
+    , dropEmptyTransition
+    , dropEmptyTransitions
+    , dropEmptyTransitionWhere
+    , dropEmptyTransitionsWhere
+    , dropHead
+    , dropLast
+    , prefixes
+    , suffixes
+
+    ) where
+
+import Prelude hiding
+    ( head, iterate, seq, tail )
+
+import Control.Applicative
+    ( ZipList (..) )
+import Control.Monad
+    ( foldM )
+import Control.Monad.Extra
+    ( allM )
+import Control.Monad.Identity
+    ( Identity (..) )
+import Data.Bifoldable
+    ( Bifoldable (..) )
+import Data.Bifunctor
+    ( Bifunctor (..) )
+import Data.Coerce
+    ( coerce )
+import Data.Function
+    ( on )
+import Data.Functor
+    ( (<&>) )
+import Data.List.NonEmpty
+    ( NonEmpty (..) )
+import Data.Sequence
+    ( Seq ((:<|), (:|>), Empty) )
+
+import qualified Data.Foldable as F
+import qualified Data.List as L
+import qualified Data.List.NonEmpty as NE
+import qualified Data.Sequence as Seq
+
+--------------------------------------------------------------------------------
+-- Types
+--------------------------------------------------------------------------------
+
+-- | The 'StateDeltaSeq' type provides a way to model an abstract sequence of
+--   state transitions, where each transition consists of an initial state
+--   value, a delta value, and a final state value.
+--
+-- Such sequences are __contiguous__, such that the final state of any given
+-- transition is the initial state of the next transition in the sequence:
+--
+-- @
+--    transition_0_1: state_0 -> delta_0_1 -> state_1
+--    transition_1_2: state_1 -> delta_1_2 -> state_2
+--    transition_2_3: state_2 -> delta_2_3 -> state_3
+--    ...
+--    transition_p_q: state_p -> delta_p_q -> state_q
+-- @
+--
+data StateDeltaSeq state delta = StateDeltaSeq
+    { head :: !state
+    , tail :: !(Seq (delta, state))
+    }
```

Comment:
Perhaps remove these `!` from here.

Reply:
However, there is also this:
https://hackage.haskell.org/package/strict-containers-0.1/docs/Data-Strict-Sequence.html

Reply:
However, it's currently difficult to experiment with changing this to [`Seq`](https://hackage.haskell.org/package/strict-containers-0.1/docs/Data-Strict-Sequence.html) from [`strict-containers`](https://hackage.haskell.org/package/strict-containers-0.1), as the version of [`cardano-base`](https://github.com/input-output-hk/cardano-base) on which we currently depend also exports a library called [`strict-containers`](https://github.com/input-output-hk/cardano-base/tree/7e3ddba98a61900181fe63cdd4c9ed9708d1a6a7/strict-containers), and Cabal (currently) doesn't make it possible to depend upon two libraries with identical names.

This naming conflict was fixed in https://github.com/input-output-hk/cardano-base/pull/308, but it'll take a while for this change to become available in our environment.

So for now, I'll shelve the experiment of changing this to the strict version, and remove the bang operators.

Reply:
Fixed in 8d5bcaecd8dae708d27ddb7aa45b77e6d570c825.